File: occasum-core/src/main/java/pw/stamina/occasum/Named.java
package pw.stamina.occasum;

import java.util.Locale;
import java.util.Objects;

public interface Named {

    String getName();

    String getId();

    //TODO Doc
    static String validateName(String name) {
        Objects.requireNonNull(name, "name is null");

        if (name.isEmpty()) {
            throw new IllegalArgumentException("name cannot be empty");
        } else if (name.contains(".")) {
            throw new IllegalArgumentException("name may not contain periods ('.')");
        }

        return name;
    }

    //TODO Doc
    static String createIdFromName(String name) {
        String id = name;

        id = id.toLowerCase(Locale.ROOT);
        id = id.replaceAll("(\\s|_)+", "_");

        return id;
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/Named.kt
package pw.stamina.occasum

import java.util.*

interface Named {

    val name: String

    val id: String

    companion object {

        //TODO Doc
        fun validateName(name: String): String {
            require(name.isNotBlank()) { "name cannot be blank" }
            require(!name.contains(".")) { "name may not contain periods ('.')" }

            return name
        }

        //TODO Doc
        fun createIdFromName(name: String): String {
            return name
                    .toLowerCase(Locale.ROOT)
                    .replace("(\\s|_)+".toRegex(), "_")
        }
    }
}
